stl/inc/algorithm

```diff
@@ -9320,6 +9333,114 @@ namespace ranges {
 #endif // _HAS_CXX20
 #endif // _HAS_CXX17
 
+template <class _RanIt, class _Pr>
+_CONSTEXPR20 void _Swap_med5_unchecked(_RanIt _First, _RanIt _Dest, _Pr _Pred) {
+    _RanIt _Guess = _First;
+    ++_Guess;
+    _RanIt _Third = _Guess;
+    ++_Third;
+    _STD _Med3_unchecked(_First, _Guess, _Third, _Pred);
+
+    _RanIt _Fourth = _Third;
+    ++_Fourth;
+    _RanIt _Fifth = _Fourth;
+    ++_Fifth;
+    if (_DEBUG_LT_PRED(_Pred, *_Fifth, *_Fourth)) {
+        swap(*_Fourth, *_Fifth); // intentional ADL
+    }
+
+    // we have _First <= _Guess (second) <= _Third and _Fourth <= _Fifth
+    if (_DEBUG_LT_PRED(_Pred, *_Fifth, *_Guess)) {
+        if (_DEBUG_LT_PRED(_Pred, *_First, *_Fifth)) {
+            _Guess = _Fifth;
+        } else {
+            _Guess = _First;
+        }
+    } else if (_DEBUG_LT_PRED(_Pred, *_Guess, *_Fourth)) {
+        if (_DEBUG_LT_PRED(_Pred, *_Third, *_Fourth)) {
+            _Guess = _Third;
+        } else {
+            _Guess = _Fourth;
+        }
+    }
+
+    swap(*_Dest, *_Guess); // intentional ADL
+}
+
+template <class _RanIt, class _Pr>
+_CONSTEXPR20 void _Nth_element_median_of_medians(_RanIt _First, _RanIt _Nth, _RanIt _Last, _Pr _Pred) {
+    using _Diff = _Iter_diff_t<_RanIt>;
+    for (;;) {
+        auto _Length = _Last - _First;
+        if (_Length <= _ISORT_MAX) {
+            _STD _Insertion_sort_unchecked(_First, _Last, _Pred); // sort any remainder
+            return;
+        }
+
+        _Diff _Median_count = _Length / static_cast<_Diff>(5);
+        auto _Median_last   = _First + _Median_count;
+        auto _Current       = _First;
+        for (auto _Median_dest = _First; _Median_dest != _Median_last; ++_Median_dest) {
+            _STD _Swap_med5_unchecked(_Current, _Median_dest, _Pred);
+            _Current += static_cast<_Diff>(5);
+        }
+
+        auto _Median_mid = _First + (_Median_count >> 1);
+        _STD _Nth_element_median_of_medians(_First, _Median_mid, _Median_last, _Pred);
+
+        // the medians before and after _Median_mid are
+        // less equal and greater equal _Median_mid, respectively.
+        // move the ones after _Median_mid near _Last and then partition
+        // the elements between the medians at front and back by the pivot _Median_mid

```

Comment:
nit - lit would be good for the comment to explain why we're moving the elements near `_Last`.

It's a bit surprising (to me, at least), so I think would be good to call out that this is an optimization (and why it is one).